File: finanze/infrastructure/client/rates/crypto/p2s_client.py
import logging

from domain.dezimal import Dezimal
from infrastructure.client.http.http_session import get_http_session


class P2SClient:
    BASE_URL = "https://api.price2sheet.com/raw"
    DEFAULT_TIMEOUT = 7

    SYMBOLS = {
        "BTC",
        "ETH",
        "LTC",
        "TRON",
        "TRX",  # alias for TRON
        "BNB",
        "USDT",
        "USDC",
        "SOL",
        "ADA",
        "DOGE",
        "DOT",
        "XRP",
        "XMR",
        "AVAX",
        "MATIC",
        "LINK",
        "ATOM",
        "UNI",
        "XLM",
        "FTM",
    }
    SYMBOL_OVERRIDES = {"TRX": "tron"}

    def __init__(self):
        self._log = logging.getLogger(__name__)
        self._timeout = self.DEFAULT_TIMEOUT
        self._session = get_http_session()

    def supports_symbol(self, symbol: str) -> bool:
        return symbol.upper() in self.SYMBOLS

    async def get_price(
        self, symbol: str, fiat_iso: str, request_timeout: int | None = None
    ) -> Dezimal:
        effective_timeout = request_timeout or self._timeout
        crypto_symbol = self.SYMBOL_OVERRIDES.get(symbol.upper(), symbol).lower()
        fiat = fiat_iso.lower()
        url = f"{self.BASE_URL}/{crypto_symbol}/{fiat}"
        raw = await self._fetch(url, effective_timeout)
        return Dezimal(raw)

    async def _fetch(self, url: str, request_timeout: int) -> str:
        response = await self._session.get(url, timeout=request_timeout)
        if response.ok:
            return await response.text()
        body = await response.text()
        self._log.error("Error Response Body:" + body)
        response.raise_for_status()
        return ""

File: mobile/src/services/client/rates/crypto/p2sClient.ts
import { Dezimal } from "@/domain/dezimal"
import { httpGetText } from "@/services/client/http/httpClient"

export class P2SClient {
  static readonly BASE_URL = "https://api.price2sheet.com/raw"
  static readonly DEFAULT_TIMEOUT = 7

  static readonly SYMBOLS = new Set<string>([
    "BTC",
    "ETH",
    "LTC",
    "TRON",
    "TRX",
    "BNB",
    "USDT",
    "USDC",
    "SOL",
    "ADA",
    "DOGE",
    "DOT",
    "XRP",
    "XMR",
    "AVAX",
    "MATIC",
    "LINK",
    "ATOM",
    "UNI",
    "XLM",
    "FTM",
  ])

  static readonly SYMBOL_OVERRIDES: Record<string, string> = {
    TRX: "tron",
  }

  supportsSymbol(symbol: string): boolean {
    return P2SClient.SYMBOLS.has(symbol.toUpperCase())
  }

  async getPrice(
    symbol: string,
    fiatIso: string,
    timeout?: number | null,
  ): Promise<Dezimal> {
    const effectiveTimeout = timeout ?? P2SClient.DEFAULT_TIMEOUT

    const override = P2SClient.SYMBOL_OVERRIDES[symbol.toUpperCase()]
    const cryptoSymbol = (override ?? symbol).toLowerCase()
    const fiat = fiatIso.toLowerCase()

    const url = `${P2SClient.BASE_URL}/${cryptoSymbol}/${fiat}`
    const { response, data } = await httpGetText(
      url,
      undefined,
      effectiveTimeout,
    )
    if (!response.ok) {
      console.error("Error Response Body:" + data)
      throw new Error(`P2S request failed: ${response.status}`)
    }

    return Dezimal.fromString(data)
  }
}
